### DISCUSSION TITLE
The relationship between firmware version and GAP SDK version

### DISCUSSION THREAD

**Discussion by SUJUSTDO** [bitcraze/discussions#1657] [Q&A]
Is there a correlation between the **firmware version of the drone** and the **gap sdk version** of gap8 in the ai-deck and the **cfclient client version**? Can I use an older firmware version with a newer sdk version?

**Comment by gemenerik** [comment]
You should generally be able to use most gap sdk versions with the AI-deck, but getting examples and CPX to work will likely take some effort.

The crazyflie-firmware and cfclient versions are more tightly coupled, but backward compatibility is typically well-maintained. However, lager mismatches in versions might result in issues. Note that cflib and cfclient are much more important to match, I would highly recommend using the same release for both.

We strive to make the latest versions of everything the best and most compatible choice. Is there a particular reason you’re looking to mix versions?

**Comment by SUJUSTDO** [comment]
I want to control the flight of the drone from the output of the AI deck. So I am using gap_sdk, cfclient and firmware respectively. The reason why I am using different versions is because when I download the ai-deck code in the docker environment, I notice that it uses the latest version of gap_sdk. whereas the local firmware version I am using is version 2022.1. I was wondering if this was possible?

**Reply by gemenerik** [reply]
I think that should be fine, just note that [CPX was been introduced after in 2022.5](https://www.bitcraze.io/2022/05/release-2022-05-with-new-aideck-functionality/) and [improved over further releases](https://www.bitcraze.io/2022/06/working-on-cpx-current-status-and-future/). You will thus have to write your own little communication protocol. This is definitely possible, I did this myself back in the day for https://github.com/tudelft/nanoflownet. Although you can definitely make something more robust, that can help you get started.

**Reply by SUJUSTDO** [reply]
Thank you for your reply, does using uart communication have any effect?

**Reply by gemenerik** [reply]
[CPX uses UART to communicate with the GAP8 via ESP32](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/#transports). So as long as you implement something good, it shouldn't matter at all. You could also communicate directly between STM32 and GAP8 over the UART, check the schematics and datasheets to identify which one. That said, it may be less effort just upgrading to the latest firmware release, depending on what is withholding you from doing that right now.
